### PR TITLE
Extracts leader attribute for subordinate units

### LINKED ISSUE #374: `unit.is_leader_from_status` reports False for all units of a subordinate charm
**Context**
python-libjuju v2.6.3
juju --version => 2.6.10-bionic-amd64 (snap)

This affects the migration of the functional test of openstack charms to the zaza testing framework, as some of them are deployed as subordinates (neutron-openvswitch and others in https://bugs.launchpad.net/charm-aodh/+bug/1828424)

**How to reproduce:**

Start with a deployed bundle including subordinate charms (e.g. neutron-openvswitch below). `juju status` reports a leader: neutron-openvswitch/0*

image

However `unit.is_leader_from_status` (https://github.com/juju/python-libjuju/blob/master/juju/unit.py#L246) reports False for all units. There is a comment in the code about returning False for subordinates on purpose. However as shown above, `juju status` does report a leader. `juju run --unit neutron-openvswitch/leader <cmd>` does work as expected as well.

Code to reproduce the issue:

```python
from juju import loop
from juju.model import Model

async def subordinate_has_no_leader():
    model = Model()
    await model.connect()

    app_name = 'neutron-openvswitch'
    app_ovs = model.applications[app_name]
    print("app: {}, units: {}".format(app_name, app_ovs.units))

    for unit in app_ovs.units:
        print("unit: {}, leader?: {}".format(unit.entity_id,
            await unit.is_leader_from_status()))

    await model.disconnect()

if __name__ == "__main__":
    loop.run(subordinate_has_no_leader())
```

which outputs the following for the model above (works fine for a principle charm):

```
$ python reproduce.py
app: neutron-openvswitch, units: [<Unit entity_id="neutron-openvswitch/1">, <Unit entity_id="neutron-openvswitch/0">]
unit: neutron-openvswitch/1, leader?: False
unit: neutron-openvswitch/0, leader?: False
```

### PR DESCRIPTION
**Pull request by tlm**
Fixes #374 by adding logic to check the subordinate status of a unit. The value of this field will dictate the traversal of the application data for finding a units leader status.

### CODE REVIEW

**Comment by tlm** [comment]
[python lib bug with is unit leader](https://trello.com/c/XGtikBMe/1008-python-lib-bug-with-is-unit-leader)

**Comment by tlm** [comment]
@hpidcock take a look. Haven't handled the case of being a subordinate more than once. Is this something we need to cover ?

File: juju/unit.py [hunk 1]
```diff
@@ -271,14 +272,32 @@ def status_history(self, num=20, utc=False):
...
             return False
 
-        if not status.applications[app].get('units'):
-            return False
 
-        if not status.applications[app]['units'].get(self.name):
+        # We will attempt to look in two places for a leader property based on
+        # if the unit is subordinate or not. These variables allow for more
+        # generic non discriminate checks
+        appData = status.applications.get(app)
```
Comment (hpidcock) [inline juju/unit.py]: app_data

Comment (tlm) [reply juju/unit.py]: go != python, Done & Done

**Comment by hpidcock** [comment]
$$merge$$

**Comment by hpidcock** [comment]
@SimonRichardson did you figure out why the merge didn't work last time?

**Comment by hpidcock** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
@tlm I know this has landed, but can we add an integration test for this?

**Pull request merged**